openshift/default-catalog-consistency/pkg/check/catalog.go

```diff
@@ -0,0 +1,118 @@
+		Name: "CatalogHasNoBundleObjectProperties",
+		Fn: func(ctx context.Context, cfg declcfg.DeclarativeConfig) error {
+			var errs []error
+			for _, bundle := range cfg.Bundles {
+				var found bool
+				for _, prop := range bundle.Properties {
+					if prop.Type == "olm.bundle.object" {
+						found = true
+					}
+				}
+				if found {
+					errs = append(errs, fmt.Errorf("bundle %q in package %q has olm.bundle.object set",
+						bundle.Name, bundle.Package))
+				}
+			}
+
+			return errors.Join(errs...)
+		},
+	}
+}
+
+// CatalogHasValidMetadataForChannelHeads checks if the channel heads have olm.csv.metadata
+// OLM Package Server (used by tooling such as the kubectl operator plugin and the OpenShift console)
+// does require these properties to be able to serve metadata about the packages in a catalog.
+//
+// In order to satisfy the needs of the package server, catalog maintainers should use this property to
+// include the CSV metadata for all bundles that are channel heads.
+// As of `opm` version 1.28.0, this property is automatically generated by when migrating SQLite
+// catalogs with `opm migrate` and when rendering SQLite catalogs and registry+v1 bundles with `opm render`.
+// Catalogs containing `olm.csv.metadata` properties must be served by `opm` binaries
+// with version at least `1.28.0`.
+func CatalogHasValidMetadataForChannelHeads() CatalogCheck {
+	return CatalogCheck{
+		Name: "CatalogHasValidMetadataForChannelHeads",
+		Fn: func(ctx context.Context, cfg declcfg.DeclarativeConfig) error {
+			var errs []error
+			packages, err := declcfg.ConvertToModel(cfg)
+			if err != nil {
+				return fmt.Errorf("unable to covert: %w", err)
+			}
+
+			for _, pkg := range packages {
+				for _, ch := range pkg.Channels {
+					channelHead, err := ch.Head()
+					if err != nil {
+						errs = append(errs, fmt.Errorf(
+							"channel %q in package %q could not determine channel head: %v",
+							ch.Name, pkg.Name, err))
+						continue
+					}
+
+					var found bool
+					for _, prop := range channelHead.Properties {
+						if prop.Type == "olm.csv.metadata" {
+							found = true
+						}
+					}
+					if !found {
+						errs = append(errs,
+							fmt.Errorf("bundle %q in package %q is missing olm.csv.metadata",
+								channelHead.Name,
+								channelHead.Package.Name))
+
+						// If we have a channel head without olm.csv.metadata we need to verify if the olm.bundle.object
+						// property is set. If not, then it means that the opm render | migrate command could not
+						// do the migration.
+						found = false
+						for _, prop := range channelHead.Properties {
+							if prop.Type == "olm.bundle.object" {
+								found = true
+							}
+						}
+
+						if !found {
+							errs = append(errs,
+								fmt.Errorf("bundle %q in package %q is a head of channel and does not either "+
+									"olm.csv.metadata or olm.bundle.object set",
+									channelHead.Name,
+									channelHead.Package.Name))
+						}
```

Comment:
I think I would drop all of this. While yes, having `olm.bundle.object` on the channel head makes it easier to migrate to `olm.csv.metadata`, I don't think our tests should be too concerned with _how_ to fix the problem. Let's just highlight _what_ the problems are.

That leaves us more open in the future to solve the "hows" in ways that don't come with built-in assumptions like needing to migrate from something that had `olm.bundle.object`.

Reply:
My motivation to add this check here is to ensure we do not fail in this scenario.
If so, we have a bug in the tool or in the pipeline, which would lead us to reach out to the author to get the solution fixed.

Currently, we have no scenario that fails within. So, I would like to keep ensuring we have no regressions. If w have we can have further data to know how to deal with.